ipc_service: Add icmsg backend

eobalski opened the pull request.

carlescufi commented on include/ipc/ipc_ringbuf.h:
```diff
@@ -0,0 +1,101 @@
+ * is encapsulated to a message with first byte reserved to contain message
+ * length. Single read allows to get the data written by the single write.
+ */
+struct ringbuf {
```
Why? we have a ring buffer implementation already?

carlescufi replied:
@nordic-krch FYI

pdunaj replied:
> Why? we have a ring buffer implementation already?

Hi @carlescufi , there are changes specific to this implementation. If you think the name is confusing with current code we can change it. The `mring` maybe (from messages ring)?

emob-nordic replied:
I just updated it with new convention. The memory access abstraction is renamed to: `icmsg_buf` (inter core messaging buffer). The backend that relies on this is renamed to `ipc_icmsg`.

carlocaione commented:
Several points:
- Add a PR description
- 2 commits are missing a commit message
- Instead of replacing the `static_vrings` sample, please add a new sample instead

emob-nordic commented:
> Several points:
> 
> * Add a PR description
> * 2 commits are missing a commit message
> * Instead of replacing the `static_vrings` sample, please add a new sample instead

I'm about to beautify this PR and make it open, that's just a first toss so you could see general idea. I'll update missing documentation and description soon.
When it comes to the sample this commit would be dropped. I used it for profiling.

Commit:
ipc_service: samples: Add sample that uses icmsg backend

Add sample to demonstrate newly introduced backend (icmsg).

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

pdunaj commented on subsys/ipc/ipc_service/backends/ipc_icmsg.c:
```diff
@@ -0,0 +1,223 @@
+
+	int ret = icmsg_buf_write(dev_data->tx_ib, msg, len);
+
+	if (ret < len) {
```
this will return true in ret is negative (will promote ret to unsigned int).

emob-nordic replied:
Thanks for spotting and providing (offline) fix for that. Everything in this PR should be ready for review now.

carlocaione commented on include/ipc/ipc_icmsg_buf.h:
```diff
@@ -0,0 +1,104 @@
+ */
+static inline struct icmsg_buf *icmsg_buf_init(void *buf, size_t blen)
+{
+	struct icmsg_buf *ib = buf, tmp = { .len = blen - sizeof(*ib) };
+
+	*ib = tmp;
```
is `tmp` really needed here?

emob-nordic replied:
Actually not, i will remove it.

carlocaione commented on subsys/ipc/ipc_service/backends/ipc_icmsg.c:
```diff
@@ -0,0 +1,229 @@
+static const uint8_t magic[4] = {0xDE, 0xAD, 0xBE, 0xEF};
+BUILD_ASSERT(sizeof(magic) <= UINT8_MAX);
+BUILD_ASSERT(CB_BUF_SIZE <= UINT8_MAX);
+
```
nit: empty line

emob-nordic replied:
I will move double empty line between everything thats static (including includes) and actual code. That's similar in `ipc_icmsg_buf.c` file. Im used to it, ii like it this way.

carlocaione commented on subsys/ipc/ipc_service/backends/ipc_icmsg.c:
```diff
@@ -0,0 +1,229 @@
+	atomic_set(&dev_data->state, ICMSG_STATE_BUSY);
+
+	int ret = mbox_init(instance);
+
```
empty line

emob-nordic replied:
It must be here, because ret is declared here. Zephyr's checkpatch will complain if i remove this line.

carlocaione replied:
can you paste the checkpath error? And please move the `ret` definition an the beginning of the function.

emob-nordic replied:
I can't trigger this locally - i don't know why. Maybe something has change since my last patches to Zephyr. Anyway i will push changes so it will be as you want.

carlocaione commented on subsys/ipc/ipc_service/backends/ipc_icmsg.c:
```diff
@@ -0,0 +1,229 @@
+
+	ret = icmsg_buf_write(dev_data->tx_ib, magic, sizeof(magic));
+	if (ret < sizeof(magic)) {
+		__ASSERT_NO_MSG(ret == sizeof(magic));
```
should this assert be outside the `if` block?

emob-nordic replied:
It won't change anything and.. i like it this way... I prefer to check returned code (with if) before i assert it. I can change it if you insist.

carlocaione commented on subsys/ipc/ipc_service/backends/ipc_icmsg.c:
```diff
@@ -0,0 +1,229 @@
+#define DT_DRV_COMPAT	zephyr_ipc_icmsg
+#define CB_BUF_SIZE	CONFIG_IPC_SERVICE_BACKEND_ICMSG_CB_BUF_SIZE
+
+static const uint8_t magic[4] = {0xDE, 0xAD, 0xBE, 0xEF};
```
ok, bikeshedding here but are you sure you want to have `0xDEADBEEF` as magic? Maybe something more custom?

emob-nordic replied:
I will change it, forgot to update it before the push.

carlocaione commented on subsys/ipc/ipc_service/backends/ipc_icmsg.c:
```diff
@@ -0,0 +1,229 @@
+#define CB_BUF_SIZE	CONFIG_IPC_SERVICE_BACKEND_ICMSG_CB_BUF_SIZE
+
+static const uint8_t magic[4] = {0xDE, 0xAD, 0xBE, 0xEF};
+BUILD_ASSERT(sizeof(magic) <= UINT8_MAX);
```
not useful. You are the one setting the `magic`

emob-nordic replied:
I forgot to put `CB_BUF_SIZE` instead of `UINT8_MAX` and then it would be useful.

carlocaione commented on subsys/ipc/ipc_service/backends/ipc_icmsg.h:
```diff
@@ -0,0 +1,13 @@
+#include <ipc/ipc_icmsg_buf.h>
+
+enum icmsg_state {
+	ICMSG_STATE_OFF = 0,
+	ICMSG_STATE_BUSY = 1,
+	ICMSG_STATE_READY = 2
```
`= 0, = 1, ...` not needed

carlocaione commented on subsys/ipc/ipc_service/lib/ipc_icmsg_buf.c:
```diff
@@ -0,0 +1,149 @@
+#include <ipc/ipc_icmsg_buf.h>
+
+
+#define cache_invalidate(region, size) /* nop */
+#define cache_flush(region, size) /* nop */
```
why is that?

emob-nordic replied:
I'm planning to add cache management later on. As you could notice read/write functions are already prepared for that. We just don't have correct calls for that in Zephyr. At least i don't see them.

carlocaione replied:
yes, we have. See `sys_cache_data_range()`

emob-nordic replied:
Pushed.

Commit:
ipc_service: Add new backend based on custom messaging buffer

This commit adds ipc backend, that relies on simple
inter core messaging buffer also added by this commit.

This backend is configurable with DT overlay. Each
ipc instance could be defined with ipc-icmsg commpatible.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

carlocaione commented on subsys/ipc/ipc_service/backends/ipc_icmsg.c:
```diff
@@ -0,0 +1,230 @@
+	}
+
+	int ret = icmsg_buf_write(dev_data->tx_ib, msg, len);
+
```
empty line

emob-nordic replied:
I can't remove this empty line. This style is according to coding style. Checkpatch will complain if i remove it.

emob-nordic commented on subsys/ipc/ipc_service/backends/Kconfig.icmsg:
```diff
@@ -0,0 +1,16 @@
+
+config IPC_SERVICE_BACKEND_ICMSG_CB_BUF_SIZE
+	int "Size of static callback buffer size"
+	range 1 255
```
Implementation limitation. Memory abstraction is written in a way that stores messages. Message is encapsulated with prefixing message len that is one byte. I didn't want to reserve two bytes for that - simply because the use case we have is mostly short messages ~10-30 Bytes. fyi rpmsg has 512 Bytes limitation.

pdunaj replied:
We have considered implementing expandable size type but decided it's overkill now.

As for reserving the value, I don't think it is needed as long as code will recompile for both cores. I understand you would like to play it safe in case somebody would like to update FW one one core only. The thing is old FW would not understand the reserved value anyway so new FW would not be able to use it. It will also not use the reserved value itself. Essentially we would need update on both cores anyway.

I don't think we need to reserve the value from protocol point of view.

hubertmis replied:
> Essentially we would need update on both cores anyway.

Fair enough

emob-nordic replied:
I saw it brought many doubts. I expanded message length to `uint16_t` so is now in `range 1 65535`

emob-nordic commented on subsys/ipc/ipc_service/lib/ipc_icmsg_buf.c:
```diff
@@ -0,0 +1,153 @@
+	return (idx >= len) ? (idx - len) : (idx);
+}
+
+int icmsg_buf_write(struct icmsg_buf *ib, const char *buf, uint8_t len)
```
I think i answered in the comment above.

emob-nordic commented on subsys/ipc/ipc_service/backends/ipc_icmsg.c:
```diff
@@ -0,0 +1,230 @@
+	return 0;
+}
+
+#define DEFINE_BACKEND_DEVICE(i)							\
```
That's not true. There are multiple places where its done 1:1 as here. Checkpatch does not trigger any warnings about that, i assume its according to the coding style.

emob-nordic commented on subsys/ipc/ipc_service/lib/ipc_icmsg_buf.c:
```diff
@@ -0,0 +1,153 @@
+#include <cache.h>
+#include <ipc/ipc_icmsg_buf.h>
+
+#define cache_invalidate(region, size) sys_cache_data_range(region, size, K_CACHE_INVD)
+#define cache_flush(region, size) sys_cache_data_range(region, size, K_CACHE_WB)
+
+#define write_barrier() __sync_synchronize()
+#define read_barrier() __sync_synchronize()
```
I will clear this up.

pdunaj commented on subsys/ipc/ipc_service/backends/ipc_icmsg.c:
```diff
@@ -0,0 +1,230 @@
+static void mbox_callback_process(struct k_work *item)
+{
+	struct backend_data_t *dev_data = CONTAINER_OF(item, struct backend_data_t, mbox_work);
+	static uint8_t cb_buffer[CB_BUF_SIZE];
```
Could be. I expect it is me - I don't like placing large buffers on the stack.

@emob-nordic since we are at it. Make this buffer word aligned.

hubertmis replied:
No strong opinion here. I think it could be more efficient to put it on stack as long as we know exactly which stack is being used. I would prefer `static` if this function was called from IRQ context, because than we don't have full control which stack is being used.

hubertmis replied:
No strong opinion here. I think it could be more efficient to put it on stack as long as we know exactly which stack is being used. I would prefer `static` if this function was called from IRQ context, because than we don't have full control which stack is being used.

emob-nordic replied:
I will follow your suggestions, remove static and align the buffer.

Commit:
tests: lib: Add ipc_icmsg_buf unit tests

This commit adds unit tests for ipc_icmsg_buf library of ipc_service.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

pdunaj commented on samples/subsys/ipc/ipc_service/icmsg/CMakeLists.txt:
```diff
@@ -0,0 +1,32 @@
+#
```
I think @emob-nordic wanted to modify the original sample but was asked to add a separate sample :)
The value the change carries is showing the bandwidth limits. We can add this code later on either by modifying the single ipc_service sample or adding a bandwidth checking specific test.

@emob-nordic If this is a controversy I propose to drop the change related to creating a sample.

hubertmis replied:
Probably my idea is not possible with Kconfig, because ipc_service backend is configured with dt. The sample would need to contain dt overrides for each backend, and potentially overrides for boards or socs. Not sure if it would be feasible to maintain. What do you think?

emob-nordic replied:
My main goal of this PR is to add backend. I don't want to block this PR because sample is too controversial. I will remove the commit that adds the sample and add it in separate PR, would this be okay @hubertmis  @carlocaione ?.

emob-nordic replied:
+ i will remove the sample commit and add unit tests for `ipc_icmsg_buf`.

hubertmis replied:
I think it's find to create a sample like the one you've already created. We can merge them later if we find it useful.
I prefer to have a separated sample than to have none. Without a sample it's hard to test changes.

carlocaione replied:
agree. Please leave the sample in place as a different one from the one testing the other backend.

emob-nordic replied:
Will bring it back then.

hubertmis commented on subsys/ipc/ipc_service/lib/ipc_icmsg_buf.c:
```diff
@@ -0,0 +1,147 @@
+int icmsg_buf_write(struct icmsg_buf *ib, const char *buf, uint8_t len)
+{
+	/* The length of buffer is immutable - avoid reloading. */
+	const uint32_t iblen = ib->len;
```
My bad, I've missed compiler memory barriers. Could you add a note to the comment that it is because of the barriers?

hubertmis commented on subsys/ipc/ipc_service/lib/ipc_icmsg_buf.c:
```diff
@@ -0,0 +1,147 @@
+
+	if (bytes_stored < mlen + sizeof(mlen)) {
+		/* Part of message not available. Should not happen. */
+		__ASSERT_NO_MSG(false);
```
Try to convince all Zephyr product makers to disable assertions in their products ;) I'm missing separated `debug_assert!()` and `assert!()` in C.

pdunaj replied:
Assert is really not substitute for error checking - if somebody uses this in production builds they have bigger problems. I won't try to convince anybody how to make products (unless they pay me to do so ;)).

Zephyr is wise enough to emit a noticeable warning message during build time that asserts are enabled.

We could remove this assert but that will make it harder to discover bugs - something in the higher layer (in the sample) would have to deal with the potential issue.
Anyway, if you think this is needed to get an approval I will not object :)

hubertmis replied:
My point was to raise awareness of this potential attack. I'm OK to merge it with it, because as you described asserts are disabled in production builds, and even if enabled DoS is not catastrophic.

carlocaione commented on subsys/ipc/ipc_service/backends/ipc_icmsg.c:
```diff
@@ -0,0 +1,230 @@
+		}
+
+		/* Reading 0 to know if there are data in the buffer. */
+		len = icmsg_buf_read(dev_data->rx_ib, NULL, 0);
+		if (len > 0) {
```
I think you can skip using `len` at all.

nordic-krch reviewed (commented):
couple of comments from review i started yesterday. Will continue today.

emob-nordic commented on subsys/ipc/ipc_service/backends/ipc_icmsg.c:
```diff
@@ -0,0 +1,230 @@
+		}
+	} else {
+		__ASSERT_NO_MSG(state == ICMSG_STATE_BUSY);
+		if (len == sizeof(magic) && !memcmp(magic, cb_buffer, len)) {
+			if (dev_data->ops && dev_data->ops->bound) {
+				dev_data->ops->bound(NULL);
+			} else {
+				__ASSERT_NO_MSG(false);
+				return;
+			}
+
+			atomic_set(&dev_data->state, ICMSG_STATE_READY);
+		} else {
+			__ASSERT_NO_MSG(false);
+			return;
+		}
```
Okay, i will redo it this time, but please don't focus on such changes. One likes doing things in X way the other in Y way. If both solutions are good and neither break anything does it matter that much? I also have some preferences in coding style and until any coding style is enforced let's not put coding style over code correctness. I don't believe its worthwhile to point out such changes in open source code.

carlocaione replied:
fair enough

emob-nordic commented on include/ipc/ipc_icmsg_buf.h:
```diff
@@ -0,0 +1,106 @@
+ *
+ * Size of the data written to the buffer is actually len + sizeof(len).
+ * First byte is always reserved to contain message length (len).
+ * Maximal length of the data written is limited to UINT8_MAX.
```
When i started the implementation we discussed what are the needs and for our use case. The outcome was we will mostly need very short messages (~<40 bytes) + the `icmsg_buf` is intendet to be small (<=2kB, max 4kB in very specific use case) . There was a decision not to lose additional byte for message len. I can see it brings more and more doubts, which surprises me as rpsmg buffer length is limited to 512 bytes, it is twice as much but still its not `uint16_t`.

I will tune it to `uint16_t`.

emob-nordic commented on include/ipc/ipc_icmsg_buf.h:
```diff
@@ -0,0 +1,106 @@
+/**
+ * @brief Read specified amount of data from the inter core messaging buffer.
+ *
+ * If len == 0, this call will return how many bytes could be read with next
```
Okay.

emob-nordic commented:
I've pushed some changes. Most important one is i removed commit with sample, at least for now. I replaced this with unit tests for `ipc_icmsg_buf`. Please have a look if you are interested (2nd commit).

@carlocaione I think i addressed all your comments.
@nordic-krch s/a except for uint16_t as for len
@rakons s/a
@hubertmis  I think we should be settled now.

I will update write and read API to uint16_t with next push, be patient.

nordic-krch commented on subsys/ipc/ipc_service/backends/ipc_icmsg.c:
```diff
@@ -0,0 +1,229 @@
+		return ret;
+	}
+
+	(void)k_work_submit(&dev_data->mbox_work);
```
is it necessary? is it to handle case when event came to mbox before it got initialized?

nordic-krch replied:
maybe we could do that only if there is pending data in the buffer?

emob-nordic replied:
For the first question - yes. I didn't want to assume if mbox event would be stored even it arrived before it was initialized.
For the second - theoretically yes, but i wanted to always schedule as first thing the work does is check if data are in the buffer. It is low cost as it happens only once, but i understand the justification - i will change that.

nordic-krch commented on subsys/ipc/ipc_service/backends/ipc_icmsg.c:
```diff
@@ -0,0 +1,229 @@
+	__ASSERT_NO_MSG(len <= CB_BUF_SIZE);
+
+	if (len == -EAGAIN) {
+		(void)k_work_submit(&dev_data->mbox_work);
```
imo this should be avoided. If callback is called there should be pending data. We can easily turn error condition into continuous polling that is not detected.

emob-nordic replied:
My idea was to react to at least try to recover from unexpected behavior. Do you suggest this should be removed and ASSERTED?

pdunaj replied:
This should not occur in any case. If it does however the best we can do is to retry.

The only case where this is a problem is when, for some reason the other core dies out before whole message is delivered (should not happen as we update index with corresponding len). If it does we will get into resubmitting loop. Maybe indeed the best thing to do, assuming this cannot happen is return and do nothing...

nordic-krch reviewed (commented):
Few comments

emob-nordic commented on include/ipc/ipc_icmsg_buf.h:
```diff
@@ -0,0 +1,106 @@
+ *
+ */
+struct icmsg_buf {
+	uint32_t len;		/* Length of data[] in bytes. */
```
I don't see any use of those flags for now. If anyone would later on expand this - i have no objections. For now this is minimal required version and i would like to avoid adding something that _could be_ useful in the future.

emob-nordic commented:
@nashif could you have a look in CI workflow? I've added some unit tests for the functionality and for some, unknown to me, reason CI is failing.

stephanosio commented:
> @nashif @stephanosio could you have a look in CI workflow? I've added some unit tests for the functionality and for some, unknown to me, reason CI is failing.

```
ERROR   - /__w/zephyr/zephyr/tests/lib/icmsg/src/main.c: Found a test that does not start with test_
ERROR   - /__w/zephyr/zephyr/tests/lib/icmsg: can't load (skipping): /__w/zephyr/zephyr/tests/lib/icmsg/src/main.c: Found a test that does not start with test_
```

stephanosio commented on tests/lib/icmsg/src/main.c:
```diff
@@ -0,0 +1,105 @@
+
+static uint8_t memory_area[212];
+
+static void icmsg_buf_ut(void)
```
```suggestion
static void test_icmsg_buf_ut(void)
```

emob-nordic replied:
Thanks a lot..

PerMac commented on tests/lib/icmsg/testcase.yaml:
```diff
@@ -0,0 +1,4 @@
+tests:
+  libraries.icmsg:
+    integration_platforms:
+      - nrf5340dk_nrf5340_cpuapp
```
Please add some meaningful way of filtering this test for twister. Integration_platforms only works for PRs. In case of a nightly build twister tries to build "everything on everything" and this test will be catch in scope for every single platform (~400) in the tree and probably fail https://docs.zephyrproject.org/latest/guides/test/twister.html#test-cases

emob-nordic replied:
This test should be platform agnostic so it probably should be run on everything possible.

emob-nordic commented on subsys/ipc/ipc_service/backends/ipc_icmsg.c:
```diff
@@ -0,0 +1,230 @@
+	atomic_t state = atomic_get(&dev_data->state);
+	int len = icmsg_buf_read(dev_data->rx_ib, cb_buffer, CB_BUF_SIZE);
+
+	__ASSERT_NO_MSG(len <= CB_BUF_SIZE);
```
Its unexpected, thats why its asserted.

pdunaj replied:
if `len` is greater then buffer we are beyond saving - internals probably trashed the stack.

Assert should be on here - we are not validating user input but making sure that internals are not broken.

emob-nordic commented on subsys/ipc/ipc_service/backends/ipc_icmsg.c:
```diff
@@ -0,0 +1,229 @@
+	}
+
+	if (state == ICMSG_STATE_READY) {
+		if (!dev_data->ops || !dev_data->ops->received) {
```
During initialization of the backend? I can't do that simply because its up to user to call `register_ept` and provide the callback.

I would like to stick to `if (!NULL)` simply because it not lead to undefined behavior when ASSERT is off. With ASSERT=n i would now return and not break anything by calling NULL pointer as a function.

emob-nordic commented on subsys/ipc/ipc_service/backends/ipc_icmsg.c:
```diff
@@ -0,0 +1,229 @@
+			return;
+		}
+
+		if (!dev_data->ops || !dev_data->ops->bound) {
```
I think i understand you know. Actually none of the callbacks can be NULL when called thats because im first registering them and later on enabling mbox. I could safely remove each assert from `mbox_callback_process`. Is that what you mean?

hubertmis replied:
I don't see anywhere in `ipc_service.h` that callbacks cannot be `NULL`. I think this check must be handled if API user is not interested in `bound()` noficiation.

emob-nordic commented on subsys/ipc/ipc_service/backends/ipc_icmsg.c:
```diff
@@ -0,0 +1,229 @@
+	struct ipc_ept *ep = (struct ipc_ept *)token;
+	int ret;
+
+	if (atomic_get(&dev_data->state) == ICMSG_STATE_BUSY) {
```
Good point, i will update it as you suggested. There is no need to distinguish those two states.

hubertmis commented on subsys/ipc/ipc_service/backends/ipc_icmsg.c:
```diff
@@ -0,0 +1,220 @@
+
+	if (!dev_data->ops->received || !dev_data->ops->bound) {
+		__ASSERT_NO_MSG(false);
+		return -EINVAL;
```
Is this really invalid? I can't find such restriction in API description.

pdunaj replied:
it would be rather pointless IPC if received is not there... 
we could question bound - but how one can know when sending is safe?

I believe it's better to not accept missing callbacks.

emob-nordic replied:
Actually @pdunaj i can see one case when the callbacks are not essential, its when one side of the communication channel does not expect to receive any data and is only focused on sending. Bound is kinda optional so i could also make it work. I will adapt it so it could work without callbacks.

hubertmis replied:
Also I can see that the reference implementation (https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c) gracefully handles when callbacks are set to `NULL`. I think we should keep consistent behavior between backend implementations.

pdunaj commented on subsys/ipc/ipc_service/lib/ipc_icmsg_buf.c:
```diff
@@ -0,0 +1,171 @@
+	uint32_t rd_idx = ib->rd_idx;
+	uint32_t wr_idx = ib->wr_idx;
+	/* Temp rd_idx used to read message len. */
+	uint32_t rd_idx_temp = rd_idx;
```
there is no point to introduce `rd_idx_temp`. simply move `bytes_stored = idx_occupied` before you read `mlen` (e.g. line 126), update `rd_idx` directly after extracting `mlen` parts and get rid of `rd_idx = idx_cut(iblen, rd_idx + sizeof(mlen));` at line 150.

Code will be simpler and more efficient.

pdunaj commented on subsys/ipc/ipc_service/lib/ipc_icmsg_buf.c:
```diff
@@ -0,0 +1,171 @@
+	/* Read message len. */
+	uint16_t mlen = ib->data[rd_idx_temp];
+	rd_idx_temp = idx_cut(iblen, wr_idx + sizeof(uint8_t));
+	mlen += ib->data[rd_idx_temp] >> 8;
```
Please use `or` instear of `add`. It cannot be slower and usually is faster. It's also what we need here. I also recommend to add brackets.

`mlen |= (ib->data[rd_idx_temp] >> 8)`

pdunaj commented on subsys/ipc/ipc_service/lib/ipc_icmsg_buf.c:
```diff
@@ -0,0 +1,171 @@
+
+	/* Read message len. */
+	uint16_t mlen = ib->data[rd_idx_temp];
+	rd_idx_temp = idx_cut(iblen, wr_idx + sizeof(uint8_t));
```
I always recommend to use size of what was read. In our case that would be 

`sizeof(ib->data[rd_idx])`

emob-nordic replied:
I introduced beautiful bug there btw. Have a look at `wr_idx` instead of `rd_idx` :hand_over_mouth:

pdunaj commented on tests/lib/icmsg/prj.conf:
```diff
@@ -0,0 +1,5 @@
+CONFIG_ZTEST=y
+
+# CONFIG_TEST_EXTRA_STACK_SIZE=1024
```
remed? `CONFIG_TEST_EXTRA_STACK_SIZE=1024`

pdunaj reviewed (changes requested):
few comments

emob-nordic commented:
+ Added commit with sample again
+ Removed assertion for callbacks existance
+ fixed a bug in ipc_icmsg_buf where `mlen` was recreated based on wr_idx + reordered code a bit to aboid `rd_idx_temp`.
+ Clean up some documentation notes (removed unnecessary lines)

hubertmis commented on subsys/ipc/ipc_service/backends/ipc_icmsg.c:
```diff
@@ -0,0 +1,220 @@
+			return;
+		}
+
+		if (dev_data->ops && dev_data->ops->received) {
```
Probably copy-paste error
```suggestion
		if (dev_data->ops && dev_data->ops->bound) {
```

emob-nordic replied:
whopsie

carlocaione commented:
Some of the CI issues should be fixed by #43903

pdunaj commented on subsys/ipc/ipc_service/lib/ipc_icmsg_buf.c:
```diff
@@ -0,0 +1,170 @@
+	ib->data[wr_idx] = (uint8_t)len;
+	wr_idx = idx_cut(iblen, wr_idx + sizeof(uint8_t));
+	ib->data[wr_idx] = (uint8_t)(len >> 8);
+	wr_idx = idx_cut(iblen, wr_idx + sizeof(uint8_t));
```
Flush cache holding parts of len
`sys_cache_data_range(&ib->data[wr_idx], sz, K_CACHE_WB);` here and after line 78

pdunaj commented on subsys/ipc/ipc_service/lib/ipc_icmsg_buf.c:
```diff
@@ -0,0 +1,170 @@
+		sys_cache_data_range(&ib->data[0], len - sz, K_CACHE_WB);
+	}
+
+	/* Update write index - make other side aware data was written. */
```
I think we could keep only one `__sync_synchronize();` just before updating wr index. (this line)
That is instead of calls in lines 86 and 92.

nordic-krch commented:
Regarding icmsg_buf. I wanted to propose an approach with non-wrapping messages. Basically, if message does not fit into buffer and would require wrapping, a padding is added (by adding 0xFF length byte) which indicates that next packet is written from the start (if there is no room at the start -ENOMEM is returned and padding is not added).

This approach slightly decreases memory utilization but simplifies implementation, reduces memory footprint (402->284 in test i run), speeds things up. Also it would allow zero-copy approach where we could get pointer to the message and free it once processed (in read case) which would allow to get rid of array allocated on the stack by upper layer.

Using 0xff as padding marker would limit max packet length to < 0xff00 which should be ok.

I sketch this proposal: https://github.com/zephyrproject-rtos/zephyr/commit/d6c53591ff8a998df0c4d4f6a8fbe3c7e245c068
With small test modification it passes the unit test.

pdunaj commented:
> Regarding icmsg_buf. I wanted to propose an approach with non-wrapping messages. Basically, if message does not fit into buffer and would require wrapping, a padding is added (by adding 0xFF length byte) which indicates that next packet is written from the start (if there is no room at the start -ENOMEM is returned and padding is not added).
> 
> This approach slightly decreases memory utilization but simplifies implementation, reduces memory footprint (402->284 in test i run), speeds things up. Also it would allow zero-copy approach where we could get pointer to the message and free it once processed (in read case) which would allow to get rid of array allocated on the stack by upper layer.
> 
> Using 0xff as padding marker would limit max packet length to < 0xff00 which should be ok.
> 
> I sketch this proposal: [d6c5359](https://github.com/zephyrproject-rtos/zephyr/commit/d6c53591ff8a998df0c4d4f6a8fbe3c7e245c068) With small test modification it passes the unit test.

That can be useful. There are tradeoffs but these might be acceptable.
Let's however first finish the PR as is. I fear that once we add another rework we will brake something and add delay.

Once we have the backend in place we could think of pros and cons of proposed change and test the performance. The only tradeoff I can think of is slight memory waste. Performance should actually improve (yes - we might delay the return of the buffer space to other core but since we get rid of one copy the overall throughput will increase). The design of len must be changed as we will no longer be able to have direct uint16_t there. But that's not a problem either as we can have variable length len.

Just please do it in subsequent PR. My whole experience is against prolonged reviews. Code gradually degrades with added feature requests as authors have enough of the changes and so rush things while reviewers also spend less time on actually analyzing the code. Let's fix what we have here and do the change iteratively.

nordic-krch commented on include/ipc/ipc_icmsg_buf.h:
```diff
@@ -0,0 +1,95 @@
+ * @param ib	A icmsg buffer to which to write.
+ * @param buf	Pointer to the data to be written to icmsg buffer.
+ * @param len	Number of bytes to be written to the icmsg buffer.
+ * @return int	Number of bytes written, negative error code on fail.
```
typically in zephyr it is described as `retval`s. See example: https://github.com/zephyrproject-rtos/zephyr/blob/c18c5cabf190496386dedceee39028f2c56e9c54/include/sys/cbprintf.h#L271

nordic-krch commented on include/ipc/ipc_icmsg_buf.h:
```diff
@@ -0,0 +1,95 @@
+ * @param buf	Pointer to the data to be written to icmsg buffer.
+ * @param len	Number of bytes to be written to the icmsg buffer.
+ * @return int	Number of bytes written, negative error code on fail.
+ *		-EINVAL, if len is bigger than the icmsg buffer can fit at all
```
why no return -ENOMEM as well. It's more or less the same but requires additional handling since different value is returned. Maybe this case could be turned into assert?

emob-nordic replied:
I want to distinguish between those two cases as one is clearly invalid use of the buffer and the second one is correct use, but the buffer is overloaded with messages.

emob-nordic replied:
Ok, i will do it as suggested.

nordic-krch commented on include/ipc/ipc_icmsg_buf.h:
```diff
@@ -0,0 +1,95 @@
+ * @return int		Bytes read, negative error code on fail.
+ *			Bytes to be read, if buf == NULL.
+ *			-EINVAL, if message can not fit in provided buf.
+ *			-EAGAIN, if not whole message is ready yet.
```
is it possible?

emob-nordic replied:
Theoretically no, but i want to leave this error as if ASSERTs are disabled and this happens somehow. I don't have any better idea than to handle this like so. I could just assert in the code `__ASSERT_NO_MSG(bytes_stored < mlen + sizeof(mlen))` but then again, with `ASSERT=n` this is will be skipped and no error will be returned. What is your suggestion? Remove this and don't check it?

nordic-krch commented on include/ipc/ipc_icmsg_buf.h:
```diff
@@ -0,0 +1,95 @@
+ * The provided buf must be big enough to store the whole message.
+ *
+ * @param ib		A icmsg buffer to which data are to be written
+ * @param buf		Data pointer to which read data will be written.
```
"If null only length of pending data is returned. " or something similar.

nordic-krch commented on subsys/ipc/ipc_service/lib/ipc_icmsg_buf.c:
```diff
@@ -0,0 +1,170 @@
+		return -EINVAL;
+	}
+
+
```
redundant line

nordic-krch commented on subsys/ipc/ipc_service/lib/ipc_icmsg_buf.c:
```diff
@@ -0,0 +1,170 @@
+	uint16_t mlen = ib->data[rd_idx];
+
+	rd_idx = idx_cut(iblen, rd_idx + sizeof(ib->data[rd_idx]));
+	mlen |= (ib->data[rd_idx] >> 8);
```
it should probably be `<< 8`

pdunaj commented on subsys/ipc/ipc_service/lib/ipc_icmsg_buf.c:
```diff
@@ -0,0 +1,171 @@
+	/* Read message len. */
+	sys_cache_data_range(&ib->data[rd_idx], sizeof(ib->data[rd_idx]), K_CACHE_INVD);
+	uint16_t mlen = ib->data[rd_idx];
+
```
excessive line - let's keep it as line 134 - if we need this line because checkpach is pita let's add similar line at line 134 for symmetry

pdunaj replied:
dont' fix it if CI is green and there is nothing better to fix. Not worth energy spent.

nordic-krch commented on samples/subsys/ipc/ipc_service/icmsg/remote/src/main.c:
```diff
@@ -0,0 +1,135 @@
+			mlen = MLEN_0;
+		}
+
+		k_busy_wait(50);
```
please add comment why it is needed.

nordic-krch commented on samples/subsys/ipc/ipc_service/icmsg/remote/src/main.c:
```diff
@@ -0,0 +1,135 @@
+
+	if (expected_message > 'z') {
+		if (stats_every++ > 50) {
+			printk("%llu\n", (1000*cnt)/(k_uptime_get_32() - start));
```
a little description of this number would be good. You may consider to use logging here instead of printk since it is in handler (but i'm not sure if that is necessary).

nordic-krch replied:
> I can not use logger here as this must be as fast as possible

To be more precise: sample is never reaching lower priority thread because of high throughput (100% cpu load) so logging would not be able to handle messages in deferred mode (and immediate mode would be heavier than printk).

emob-nordic replied:
I will add this.

nordic-krch commented on samples/subsys/ipc/ipc_service/icmsg/remote/src/main.c:
```diff
@@ -0,0 +1,135 @@
+			stats_every = 0;
+		}
+		if (err) {
+			printk("!!\n");
```
Use more detailed error message.

nordic-krch reviewed (commented):
I would like to see `icmsg` and `static_vrings` samples merged. The point of using `ipc_service` is to show that underneath various backends can be used. Adding separate samples indicates something opposite. If `icmsg` does not support more endpoints then sample may just abort second thread when endpoint register fails. Having performance showcase in the sample would also be vital because that would allow comparing of backends.
Given that, i don't want to block this PR and sample merging can be done separately.

emob-nordic commented:
@nordic-krch  @carlocaione  Can as ask you for last round? Reviews are dismissed after i pushed changes with comments.

emob-nordic commented:
@carlescufi please have a look and merge if possible.

stephanosio commented on samples/subsys/ipc/ipc_service/icmsg/remote/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+#
+# Copyright (c) 2022 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
```
This is not allowed upstream.

@emob-nordic Can you please send a PR to fix this since you have signed off on this?

emob-nordic replied:
Will do by the end of week.

stephanosio replied:
Thanks!

emob-nordic replied:
https://github.com/zephyrproject-rtos/zephyr/pull/54491
